esp-hal/src/dma/sdio.rs

```diff
@@ -0,0 +1,468 @@
+        // hardware should trigger an interrupt request.
+        self.set_suc_eof(set_eof);
+    }
+
+    /// Set the size of the buffer. See [DmaDescriptorFlags::size].
+    pub fn set_size(&self, len: usize) {
+        self.flags.set_size(len as u16)
+    }
+
+    /// Set the length of the descriptor. See [DmaDescriptorFlags::length].
+    pub fn set_length(&self, len: usize) {
+        self.flags.set_length(len as u16)
+    }
+
+    /// Returns the size of the buffer. See [DmaDescriptorFlags::size].
+    pub fn size(&self) -> usize {
+        self.flags.size() as usize
+    }
+
+    /// Returns the length of the descriptor. See [DmaDescriptorFlags::length].
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        self.flags.length() as usize
+    }
+
+    /// Set the suc_eof bit. See [DmaDescriptorFlags::suc_eof].
+    pub fn set_suc_eof(&self, suc_eof: bool) {
+        self.flags.set_suc_eof(suc_eof)
+    }
+
+    /// Returns the suc_eof bit. See [DmaDescriptorFlags::suc_eof].
+    pub fn suc_eof(&self) -> bool {
+        self.flags.suc_eof()
+    }
+
+    /// Set the owner. See [DmaDescriptorFlags::owner].
+    pub fn set_owner(&self, owner: Owner) {
+        self.flags.set_owner(owner)
+    }
+
+    /// Returns the owner. See [DmaDescriptorFlags::owner].
+    pub fn owner(&self) -> Owner {
+        self.flags.owner()
+    }
+}
+
+// The pointers in the descriptor can be Sent.
+// Marking this Send also allows DmaBuffer implementations to automatically be
+// Send (where the compiler sees fit).
+unsafe impl Send for DmaDescriptor {}
+
+#[cfg(feature = "defmt")]
+impl defmt::Format for DmaDescriptor {
+    fn format(&self, fmt: defmt::Formatter<'_>) {
+        defmt::write!(
+            fmt,
+            "flags: {}, buffer: {}, next: {}",
+            self.flags.descriptor_flags(),
+            self.buffer.load(Ordering::Acquire),
+            self.next.load(Ordering::Acquire),
+        );
+    }
+}
+
+/// Turns the potentially oversized static `u32`` array reference into a
+/// correctly sized `u8` one
+#[doc(hidden)]
+#[macro_export]
+macro_rules! sdio_as_mut_byte_array {
+    ($name:expr, $size:expr) => {
+        const _: () = assert!($size < $name.len(), "invalid SDIO DMA buffer size");
+        unsafe { &mut $name.as_slice_mut()[..$size] }
+    };
+}
+pub use sdio_as_mut_byte_array as as_mut_byte_array;
+
+/// Declares a SDIO DMA buffer with a specific size, aligned to 4 bytes
+#[doc(hidden)]
+#[macro_export]
+macro_rules! declare_aligned_sdio_dma_buffer {
+    ($name:ident, $size:expr) => {
+        // ESP32 requires word alignment for DMA buffers.
+        // ESP32-S2 technically supports byte-aligned DMA buffers, but the
+        // transfer ends up writing out of bounds.
+        // if the buffer's length is 2 or 3 (mod 4).
+        const SIZE: usize = $size + ($size % 4);
+        static mut $name: [u8; SIZE] = [0; SIZE];
```

Comment:
This is still a `static mut`, and I plan to create a new-type to wrap it. Probably something wrapping an equivalent buffer of `AtomicU8` with the relevant traits implemented (e.g. `copy_from_slice`, `buf[i] = val;`, `IterMut`, etc.). Might be overkill, but I'd like to avoid introducing any more `static mut` in code I contribute.

Reply:
Right, I wasn't sure how important the macros were to the surrounding machinery in the `dma` module. I was trying to pull in the minimal amount of code to make the necessary changes.

I'll work on following the ideas in #2993 when iterating on this code.